### ISSUE TITLE
Caching node (memoization)

### ISSUE THREAD

**Issue by zopieux** [chaiNNer-org/chaiNNer#1466]
**Motivation**
A typical use-case of chaiNNer that I'm sure many people have is to do an expensive operation (e.g. upscale image), then proceed with further cheap processing using that expensive output as input. One needs to iterate a lot of that cheap processing. It would be great if chaiNNer could expose a caching mechanism so the expensive stuff is only done once and doesn't require manually saving/loading from disk. I propose two approaches below.

**Description**
1. One option is to introduce a *Cache node* which is basically a passthrough with state (memory and/or disk). I don't know if the current architecture can detect dirty inputs for any stage – looks like at least the *Image preview* node has a concept of "outdated" that would apply here.
2. Another (better IMO) option is to automatically cache all operations (for nodes where it makes sense) behind the scene instead of having to create cache nodes manually.

**Alternatives**
Today I need to rely on a clunky setup where I have a *Save image* node and a *Load image* node, depending on whether I need to refresh the output or just read it. This is a huge pain when there are a few such cacheable operations, especially in the absence of i/o selectors (#1465).

**Comment by joeyballentine** [comment]
We actually were planning on implementing that second option at some point, but a cache node sounds like an even better idea

**Comment by zopieux** [comment]
Cool! As I mentionned I think the user shouldn't be in the business of manually adding such node, because ideally *the default* should be for the app to be smart and detect changes. Of course this is likely more difficult to imlement correctly compared to a simple caching node.

**Comment by joeyballentine** [comment]
Yeah i think the issue there is how we define what needs to be cached. We obviously don't want to cache _evrrything_, so we'd need to specify what things would be worth caching (upscaling nodes being one of those). I do think we should still have a cache node though, since there may be times where a user is working with incredibly large images on a low end machine, and even simple nodes take a while to process, and they'd benefit from adding a big of caching around. @RunDevelopment probably has a better idea of what would be the best approach than me though
